Paired-key Crypto Scheme

drskalman opened the pull request:
BEEFY needs two cryptographic keys at the same time. Validators should sign BEEFY payload using both ECDSA and BLS key. The network will gossip a payload which contains a valid ECDSA key. The prover nodes aggregate the BLS keys if aggregation fails to verifies the validator which provided a valid ECDSA signature but an invalid BLS signature is subject to slashing. 

As such BEEFY session should be initiated with both key. Currently there is no straight forward way of doing so, beside having a session with RuntimeApp corresponding to a crypto scheme contains both keys. 

This pull request implement a generic `paired_crypto` scheme as well as implementing it for (ECDSA, BLS) pair.

Polkadot companion: (*if applicable*)

Cumulus companion: (*if applicable*)

# Checklist

- [x] My PR includes a detailed description as outlined in the "Description" section above
- [ ] My PR follows the [labeling requirements](https://github.com/paritytech/substrate/blob/master/docs/CONTRIBUTING.md#merge-process) of this project (at minimum one label for each `A`, `B`, `C` and `D` required)
- [ ] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] If this PR alters any external APIs or interfaces used by Polkadot, the corresponding Polkadot PR is ready as well as the corresponding Cumulus PR (optional)


✄ -----------------------------------------------------------------------------

drskalman committed:
First definition for pair public keys

drskalman committed:
Two example of implementation of pair for demonestration

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3327400

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3327386

drskalman commented on primitives/core/src/paired_crypto.rs:
```diff
@@ -0,0 +1,109 @@
+/// A public key.
+#[derive(Clone, Encode, Decode, MaxEncodedLen, TypeInfo, PartialEq, Eq, PartialOrd, Ord)]
+#[scale_info(skip_type_params(T))]
+pub struct PublicWithInner<LeftPublic: PublicKeyBound, RightPublic: PublicKeyBound, const LEFT_PLUS_RIGHT_SIZE: usize> {
+    inner: [u8; LEFT_PLUS_RIGHT_SIZE],
+ 	_phantom: PhantomData<fn() -> (LeftPublic, RightPublic)>,
+}
+
+pub struct Public<LeftPublic: PublicKeyBound, RightPublic: PublicKeyBound> {
+    left_public : LeftPublic,
+    right_public: RightPublic,
+}
```
@davxy : could you tell me which way of implementing 'paired_crypto` do you prefer? In my opinion the second is more clean and more expressive but it results in copying during `tryForm<[u8]>` and `AsMut<[u8]>` operations. On the other hand, if we go with the inner approach, it could be that when we try to use the subkeys, deserializing `inner` results in copying as well as we don't have control on the implementation of subkey types.

davxy replied:
IMO the second representation is better (as you said).
In this way we abstract the internal representation of the components.

It'd be interesting to try to generalize to `Tuples` (using https://docs.rs/impl-trait-for-tuples/0.2.2/impl_trait_for_tuples/) instead of defining the single types.
 In this way we'd have the implementation of Public, Pair, etc for any tuple of types implementing Public, Pair, etc...

drskalman committed:
- implement paired crypto `Public` as tuple of two `Public`s
- unsucessfuly try to implement ByteArray for paired crypto Public

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3379232

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3379218

drskalman commented on primitives/core/src/paired_crypto.rs:
```diff
@@ -0,0 +1,124 @@
+    }
+}
+
+// impl<LeftPublic: PublicKeyBound, RightPublic: PublicKeyBound, const LEFT_LEN: usize, const RIGHT_LEN: usize> AsMut<[u8]> for Public<LeftPublic, RightPublic, LEFT_LEN, RIGHT_LEN> {
+//  	fn as_mut(&mut self) -> &mut [u8] {
+//  	 &mut [self.0.as_mut(), self.1.as_mut()].concat()
+//  	}
+// }
```
@davxy it seems that the second representation `(LeftPublic: PublicBound, RightPublic: PublicBound)` does not work because we can not implement `AsRef<u8>` and `AsMut<u8>` as we don't have a consecutive representation of the sub-elements as [these people suggests](https://users.rust-lang.org/t/how-to-implement-the-trait-std-asref-u8-for-generic-structure/33554). Should I go back to the first solution:
```
pub struct PublicWithInner<LeftPublic: PublicKeyBound, RightPublic: PublicKeyBound, const LEFT_PLUS_RIGHT_SIZE: usize> {
    inner: [u8; LEFT_PLUS_RIGHT_SIZE],
 	_phantom: PhantomData<fn() -> (LeftPublic, RightPublic)>,
}
```

or should I keep both representations at the same time:
```
pub struct Public<LeftPublic: PublicKeyBound, RightPublic: PublicKeyBound, const LEFT_PLUS_RIGHT_SIZE: usize> {
    left_public : LeftPublic,
    right_public: RightPublic,
    inner: [u8; LEFT_PLUS_RIGHT_SIZE],

}
```
or you have a solution to building non-ephemeral slice in a function that works here?

🙏

drskalman committed:
keep both public key object and their continous serialization in paired crypto object in favor of avoiding copy

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3420187

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3420173

drskalman committed:
implement PassBy and From<Pair> for paired_crypto

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3421566

drskalman committed:
implement rest of aux traits for `paired_crypto::Public`
implement some of aux traits for `paired_crypto::Signature`

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3427364

drskalman committed:
Attempt to implement trait `Pair` for `pair_cyrpto::Pair`

drskalman committed:
- Implement trait `Pair` for `paired_crypto::Pair`
- Implement a pair of seeds for paired crypto scheme.

drskalman committed:
implement sgin and verify for

drskalman committed:
Actually implementing `paired_crypto::{Pair, Public, Signatrue}` for paired (ECDSA, BLS377) crypto

drskalman committed:
Implement and pass all test for `paired_crypto`

drskalman committed:
- move to signle seed for both schemes in `primitives/core/src/paired_crypto.rs`
- put serialize and descerialze under `serde` feature instead of std in `primitives/core/src/bls.rs`
- fix documentation in `primitives/core/src/bls.rs`

drskalman committed:
cargo fmt

drskalman committed:
Merge remote-tracking branch 'origin/master' into skalman-paired-scheme-crypto
